modules/build/src/main/scala/scala/build/CrossSources.scala

```diff
@@ -305,4 +317,43 @@ object CrossSources {
       .left.map(CompositeBuildException(_))
       .map(_.flatten)
 
+  /** Filters out the sources from the input sequence based on the provided 'exclude' patterns. The
+    * exclude patterns can be absolute paths, relative paths, or glob patterns.
+    *
+    * @throws BuildException
+    *   If multiple 'exclude' patterns are defined across the input sources.
+    */
+  private def excludeSources(
+    sources: Seq[PreprocessedSource],
+    workspaceDir: os.Path
+  ): Either[BuildException, Seq[PreprocessedSource]] = either {
+
+    val exclude: Either[BuildException, Seq[Positioned[String]]] =
+      sources.flatMap(_.options).map(_.internal.exclude).filter(_.nonEmpty) match {
+        case Nil      => Right(Nil)
+        case Seq(one) => Right(one)
+        case excludes => Left(new MultipleExcludeError(excludes.flatMap(_.flatMap(_.positions))))
+      }
+    val excludePatterns = value(exclude).map(_.value).flatMap { p =>
+      val maybeRelPath = Try(os.RelPath(p)).toOption
+      maybeRelPath match {
+        case Some(relPath) if os.isFile(workspaceDir / relPath) =>
+          Seq(p, (workspaceDir / relPath).toString) // exclude relative file paths
+        case Some(relPath) if os.isDir(workspaceDir / relPath) =>
+          Seq(p, s"${workspaceDir / relPath}${File.separator}*") // exclude relative directory paths
```

Comment:
A little more readable IMO
```suggestion
          Seq(p, (workspaceDir / relPath / "*").toString) // exclude relative directory paths
```